### ISSUE TITLE
Feature: Improve PowerShell elevation syntax with a wrapper function script [Help Wanted]

### ISSUE THREAD

**Issue by gerardog** [gerardog/gsudo#39]
For context: when writing this, `gsudo` is invoked as any other console `.EXE` app from PowerShell. This means the parsing/quote escaping is not ideal and [this rules](https://github.com/gerardog/gsudo/blob/af68b2fc33c80bcc018edecbdd26ad6c21d50f06/README.md#usage-from-powershell--powershell-core) must be followed.

Looking forward to implement an `invoke-gsudo' function for PowerShell and I would like to hear opinions from people with more PowerShell experience than me.

- This function would be a wrapper of `gsudo.exe` that would make it feel more `PowerShell` native.
- The function name: What would be the best name for it? I bet people would throw me stones if It doesnt respect the `verb-noun` form. Ideally it should NOT be the same as in PowerShell/PowerShell#11343 which is hard to know since that one isn't defined yet either. (reason: to avoid all flows to break when that one is released). From now on I would just say `invoke-gsudo` as an alias for `to be defined function name`. Also, maybe it would be better to leave any alias definitions to the end user.
- The deployment model: I think I figured out this one: By creating a `Invoke-gsudo.ps1` file in the PATH (e.g. gsudo folder) would be enough. The function should be deployed by the 3 installers (scoop/choco/manual `.ps1`)
- Input command parsing: Ideally one would just prepend `invoke-gsudo` without special quoting rules, but is that doable? Best way to get variable substitution? Would the PS-Remoting model work for gsudo? 
  For example, this difference is unwanted: (related #38)
  ```
   PS> echo "abc def"
   abc def
   PS> gsudo echo "abc def"
   abc
   def
  ```

- Output result marshalling: Since marshaling is impossible to avoid, this could be like: The elevated instance serializes the result instead of `.ToString()` it, stream (StdIn/Out) and non-elevated deserialize. 

Reason I wrote this is here is because I prefer to gather feedback very early on. I don't want to invest  time just to learn (after releasing) that I reinvented a wheel already available for free, in any of these areas.

**Comment by oising** [comment]
There isn't really a good reason to write a native (i.e. C# compiled) wrapper cmdlet as far as I can see? If you want to marshal parameters using powershell's binder to gsudo's, a wrapper function would do the trick. Or perhaps you're not aware that cmdlet infers compiled code, and function infers script?

**Comment by gerardog** [comment]
Oh, I didn't mean a native cmdlet, but a function script.
My intent is just to add one `Invoke-gsudo.ps1` file side by side with `gsudo.exe`, both in the PATH.
Then you could `Invoke-gsudo -ScriptBlock { Do-ElevatedStuff }` or whatever syntax is more welcomed for PS developers than the current (quote-escaping-hell) syntax...

**Comment by gerardog** [comment]
Oh, I didn't mean a native c# cmdlet, but a function script.
My intent is just to add one `Invoke-gsudo.ps1` file side by side with `gsudo.exe`, both in the PATH.
Then you could `Invoke-gsudo -ScriptBlock { Do-ElevatedStuff }` or whatever syntax is more welcomed for PS developers than the current (quote-escaping-hell) syntax...

**Comment by oising** [comment]
I'd be happy to contribute, but that said, I haven't played with your magical powershell elevation. I suspect it secretly sends the work to an out of process instance of powershell. Am I right? Trying to pretend that a scriptblock is in the same lexical scope as the caller might lead to all kinds of weird situations as the variables wouldn't exist in that other runspace's sessionstate. Or are you doing something sneakier?

**Comment by gerardog** [comment]
That's right. gsudo.exe does [very little](https://github.com/gerardog/gsudo/blob/d1b4abb3d069aaa1b5f06e6f9ecb7a1a83f689a4/src/gsudo/Helpers/ArgumentsHelper.cs#L36) to support elevating PS commands. The trick is that it is the user's responsibility to generate a string with a valid PS command (hence the quoting-hell) and then, when gsudo detects it's being called from PS, it elevates the same PowerShell.exe with `-c {command}`.

This can be seen with the --debug parameter:
```
PS $ gsudo "(Get-FileHash \""$file\"" -Algorithm $algorithm).Hash"
``` 
... executes `gsudo "(Get-FileHash \"C:\test\test.txt\" -Algorithm MD5).Hash"` which then calls pwsh again:
```
Debug: Command to run: "C:\Users\ggrignoli\scoop\apps\pwsh\current\pwsh.exe" -NoLogo -NoProfile -Command "(Get-FileHash \"C:\test\test.txt\" -Algorithm MD5).Hash"
(...)
79F1F6A36DBEBF6888E9E1717766F5A6
```
In that model, the user should be aware that there is no object marshalling, only strings in and out. In the example you can capture the result because `.Hash` returns a simple string, but good luck if it were a complex object.

So, what's next? The point of this issue is that I am open to suggestions and gather info on how far we can go.

>  Trying to pretend that a scriptblock is in the same lexical scope 

It should be clear from the ground up that its not. There is a process-hop between elevated an unelevated instances so there will be serializing and deserializing of objects. 

What is PS-Remoting doing with both local and remote scopes? Can we mimic that? 

Take a look at [this](https://stackoverflow.com/a/60583163/97471), it serializes a scriptblock while allowing the `$using:LocalVariable` syntax: it substitutes from $using:variableA to:  $(Deserialize('serialized value of variableA')).

i.e. `Write-Host $a` becomes
`Write-Host $([System.Management.Automation.PSSerializer]::Deserialize('&lt;Objs Version="1.1.0.1" xmlns="http://schemas.microsoft.com/powershell
/2004/04"&gt;_x000D__x000A_  &lt;S&gt;HELLO&lt;/S&gt;_x000D__x000A_&lt;/Objs&gt;'))`

AFAIK this would allow this syntax: `Elevate-Command -ScriptBlock { $using:localVar + $remoteVar }`, unless I am missing something.

Serialization will produce huge strings, but we can use `pwsh -c -` and then send the scriptblock via StdIn. 
The output should be serialized before going to StdOut, then deserialized as well.

**Comment by gerardog** [comment]
e.g. [Elevate-Command.ps1](https://gist.github.com/gerardog/d5e5e539d3e00407c61c59135aa145d2)

```
$a = 10
Elevate-Command -ScriptBlock { $b = $using:a ; $b = $b+1 ; $b }
```
returns `11`!

**Comment by gerardog** [comment]
New version ([same link](https://gist.github.com/gerardog/d5e5e539d3e00407c61c59135aa145d2)), supports `-ArgumentList` and pipe input is received via $InputObject.

``` powershell
$local = "outside scope"

"TestString" | Elevate-Command -ScriptBlock {
    $local = "inner scope"
    Write-Host "Am I elevated? $([bool](([System.Security.Principal.WindowsIdentity]::GetCurrent()).groups -match 'S-1-5-32-544'))"
    Write-Host "Inner scope value: $local"
    Write-Host "Outer scope value: $using:local"
    Write-Host "Received Args: $args"
    Write-Host "Pipeline Input: $Input"
} -ArgumentList 1, 2, 3

```
returns
```
Am I elevated? True
Inner scope value: inner scope
Outer scope value: outside scope
Received Args: 1 2 3
Pipeline Input: TestString
```
I've tried to make `Elevate-Command` equivalent to `Invoke-Command`, except that the latter does not supports `$using:` syntax on `localhost`.
Does it sound practical? What could be improved? Opinions?

Now I've been looking at [`Ìnvoke-Expression`](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/invoke-expression?view=powershell-7)... I am the exploring the idea of providing an `Elevate-Expression` function. I guess it will look similar to the same as `gsudo` right now, but with a few differences:
- Quoting rules would be exactly PS rules, not subtle differences like " => \""
- Support `$using:` which will work by serializing the object using PSSerializer.
 
Any feedback at this early stage is highly welcome.

**Comment by gerardog** [comment]
New version ([same link](https://gist.github.com/gerardog/d5e5e539d3e00407c61c59135aa145d2)), supports `-ArgumentList` and pipe input is received via $InputObject.

``` powershell
$local = "outside scope"

"TestString" | Elevate-Command -ScriptBlock {
    $local = "inner scope"
    Write-Host "Am I elevated? $([bool](([System.Security.Principal.WindowsIdentity]::GetCurrent()).groups -match 'S-1-5-32-544'))"
    Write-Host "Inner scope value: $local"
    Write-Host "Outer scope value: $using:local"
    Write-Host "Received Args: $args"
    Write-Host "Pipeline Input: $Input"
} -ArgumentList 1, 2, 3

```
returns
```
Am I elevated? True
Inner scope value: inner scope
Outer scope value: outside scope
Received Args: 1 2 3
Pipeline Input: TestString
```
I've tried to make `Elevate-Command` equivalent to `Invoke-Command`, except that the latter does not supports `$using:` syntax on `localhost`.
Does it sound practical? What could be improved? Opinions?

Now I've been looking at [`Ìnvoke-Expression`](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/invoke-expression?view=powershell-7)... I am the exploring the idea of providing an `Elevate-Expression` function. I guess it will look similar to the same as `gsudo` right now, but with a few differences:
- Quoting rules would be exactly PS rules, not subtle differences like " => \""
- Support `$using:` which will work by serializing the object using PSSerializer.
 
Any feedback at this early stage is highly welcome.

**Comment by gerardog** [comment]
New version ([same link](https://gist.github.com/gerardog/d5e5e539d3e00407c61c59135aa145d2)), supports `-ArgumentList` and pipe input is received via $InputObject.

``` powershell
$local = "outside scope"

"TestString" | Elevate-Command -ScriptBlock {
    $local = "inner scope"
    Write-Host "Am I elevated? $([bool](([System.Security.Principal.WindowsIdentity]::GetCurrent()).groups -match 'S-1-5-32-544'))"
    Write-Host "Inner scope value: $local"
    Write-Host "Outer scope value: $using:local"
    Write-Host "Received Args: $args"
    Write-Host "Pipeline Input: $Input"
} -ArgumentList 1, 2, 3

```
returns
```
Am I elevated? True
Inner scope value: inner scope
Outer scope value: outside scope
Received Args: 1 2 3
Pipeline Input: TestString
```
I've tried to make `Elevate-Command` equivalent to `Invoke-Command`, except that the latter does not supports `$using:` syntax on `localhost`.
Does it sound practical? What could be improved? Opinions?

Now I've been looking at [`Ìnvoke-Expression`](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/invoke-expression?view=powershell-7)... I am the exploring the idea of providing an `Elevate-Expression` function. I guess it will look similar to the same as `gsudo` right now, but with a few differences:
- Quoting rules would be exactly PS rules, not subtle differences like " => \""
- Support `$using:` which will work by serializing the object using PSSerializer.
 
Any feedback at this early stage is highly welcome.

**Comment by gerardog** [comment]
New version ([same link](https://gist.github.com/gerardog/d5e5e539d3e00407c61c59135aa145d2)), supports `-ArgumentList` and pipe input is received via $Input.

``` powershell
$local = "outside scope"

"TestString" | Elevate-Command -ScriptBlock {
    $local = "inner scope"
    Write-Host "Am I elevated? $([bool](([System.Security.Principal.WindowsIdentity]::GetCurrent()).groups -match 'S-1-5-32-544'))"
    Write-Host "Inner scope value: $local"
    Write-Host "Outer scope value: $using:local"
    Write-Host "Received Args: $args"
    Write-Host "Pipeline Input: $Input"
} -ArgumentList 1, 2, 3

```
returns
```
Am I elevated? True
Inner scope value: inner scope
Outer scope value: outside scope
Received Args: 1 2 3
Pipeline Input: TestString
```
I've tried to make `Elevate-Command` equivalent to `Invoke-Command`, except that the latter does not supports `$using:` syntax on `localhost`.
Does it sound practical? What could be improved? Opinions?

Now I've been looking at [`Ìnvoke-Expression`](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/invoke-expression?view=powershell-7)... I am the exploring the idea of providing an `Elevate-Expression` function. I guess it will look similar to the same as `gsudo` right now, but with a few differences:
- Quoting rules would be exactly PS rules, not subtle differences like " => \""
- Support `$using:` which will work by serializing the object using PSSerializer.
 
Any feedback at this early stage is highly welcome.

**Comment by gerardog** [comment]
New version ([same link](https://gist.github.com/gerardog/d5e5e539d3e00407c61c59135aa145d2)), supports `-ArgumentList` and pipe input is received via $Input.

``` powershell
$local = "outside scope"

"TestString" | Elevate-Command -ScriptBlock {
    $local = "inner scope"
    Write-Host "Am I elevated? $([bool](([System.Security.Principal.WindowsIdentity]::GetCurrent()).groups -match 'S-1-5-32-544'))"
    Write-Host "Inner scope value: $local"
    Write-Host "Outer scope value: $using:local"
    Write-Host "Received Args: $args"
    Write-Host "Pipeline Input: $Input"
} -ArgumentList 1, 2, 3

```
returns
```
Am I elevated? True
Inner scope value: inner scope
Outer scope value: outside scope
Received Args: 1 2 3
Pipeline Input: TestString
```
I've tried to make `Elevate-Command` equivalent to `Invoke-Command`, except that the latter does not supports `$using:` syntax on `localhost`.
Does it sound practical? What could be improved? Opinions?

Now I've been looking at [`Ìnvoke-Expression`](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/invoke-expression?view=powershell-7)... I am the exploring the idea of providing an `Elevate-Expression` function. I guess it will look similar to the same as `gsudo` right now, but with a few differences:
- Quoting rules would be exactly PS rules, not subtle differences like " => \""
- Support `$using:` which will work by serializing the object using PSSerializer.
 
Any feedback at this early stage is highly welcome.

**Comment by oising** [comment]
So one thing you could do is to create a proxy command for Invoke-Expression to add an -Elevate parameter. Inventing new verbs (elevate) is frowned upon in the powershell world. You can have powershell automagically generate a wrapper function and pipe it to the clipboard like so:

```powershell
[System.Management.Automation.ProxyCommand]::Create((gcm invoke-expression)) | clip
```
You end up with a function body (which you would wrap in a `function Invoke-Expression { ... }` block.) Functions have precedence over native Cmdlets. 

```powershell
[CmdletBinding(HelpUri='https://go.microsoft.com/fwlink/?LinkID=2097030')]
param(
    [Parameter(Mandatory=$true, Position=0, ValueFromPipeline=$true)]
    [string]
    ${Command})

begin
{
    try {
        $outBuffer = $null
        if ($PSBoundParameters.TryGetValue('OutBuffer', [ref]$outBuffer))
        {
            $PSBoundParameters['OutBuffer'] = 1
        }

        $wrappedCmd = $ExecutionContext.InvokeCommand.GetCommand('Microsoft.PowerShell.Utility\Invoke-Expression', [System.Management.Automation.CommandTypes]::Cmdlet)
        $scriptCmd = {& $wrappedCmd @PSBoundParameters }

        $steppablePipeline = $scriptCmd.GetSteppablePipeline($myInvocation.CommandOrigin)
        $steppablePipeline.Begin($PSCmdlet)
    } catch {
        throw
    }
}

process
{
    try {
        $steppablePipeline.Process($_)
    } catch {
        throw
    }
}

end
{
    try {
        $steppablePipeline.End()
    } catch {
        throw
    }
}
<#

.ForwardHelpTargetName Microsoft.PowerShell.Utility\Invoke-Expression
.ForwardHelpCategory Cmdlet

#>
```
You would then add a new switch parameter to the param block, `[switch]$Elevate`. This is a reasonable way to add new functionality to an out of the box cmdlet. But given that you're pushing the work to gsudo, you'd rip out the steppable pipeline stuff and replace with your own marshalling script. Just an idea.

**Comment by oising** [comment]
btw, powershell.exe and pwsh.exe can _automatically_ deserialize psserialized objects from a native command (e.g. cmd batch, exe or other out of proc process) if the stdin stream is prefixed with a special marker sequence. Check this out:

Create a batch file `foo.cmd`  (escaping < and > with ^)
```cmd
@echo #^< CLIXML
@echo ^<Objs Version="1.1.0.1" xmlns="http://schemas.microsoft.com/powershell/2004/04"^>^<I32^>42^</I32^>^</Objs^>
```
Now, pipe this command to foreach %
```powershell
.\foo.cmd | % { $_ }
42
```
Magic!

**Comment by oising** [comment]
Oh this could also help - you can get powershell to serialize AND add the magic marker itself like this:

```cmd
pwsh -noprofile -outputformat xml -command { gi c:\windows }
#< CLIXML
<Objs ... >
```
Example rehydrating in another pwsh process:

```cmd
pwsh -noprofile -outputformat xml -command { gi c:\windows } | pwsh -noprofile -command { $input | % { $_ } }
```
```text
Directory: C:\

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d----          2020-06-15  2:58 PM                Windows
```

**Comment by majkinetor** [comment]
> I bet people would throw me stones if It doesnt respect the verb-noun form. 

The correct way is to name it in verb-noun (Invoke-Gsudo) manner and set alias to whatever (gsudo).

**Comment by nickcox** [comment]
For the issue around quoting rules, you can access the invoked command with `$MyInvocation`. This seems to work for me so far:

```powershell
function Invoke-Gsudo {
  gsudo.exe ($MyInvocation.Line -replace "^$($MyInvocation.InvocationName)\s+" -replace '"','""')
}

❯ set-alias sudo Invoke-Gsudo
❯ sudo echo "abc def"
abc def

❯ _
```

**Comment by mattcargile** [comment]
The `-EncodedCommand` parameter seems helpful here too. I use it in this function I stole from so and so in order to run `psexc` , sysinternals tool.

Here is the applicable piece of code which appears in the `help` for `pwsh` and `powershell` , _i think_.

**Comment by mattcargile** [comment]
I was playing around with the new build. I couldn't understand the piped `$InputObject` piece. When I removed it, code like the below worked.

`Invoke-gsudo -ScriptBlock { Get-Content .\readline.ps1 } -NoElevate`

https://github.com/gerardog/gsudo/blob/e9ad64b9932f0e444b1604b7d56af0bc2bddae82/src/gsudo.extras/Invoke-gsudo.ps1#L110

Some of this made feed into the `$Input` variable not being defined. Should `$Input` be `$InputObject`?

https://github.com/gerardog/gsudo/blob/e9ad64b9932f0e444b1604b7d56af0bc2bddae82/src/gsudo.extras/Invoke-gsudo.ps1#L95

When I changed the above line to `$InputObject` , I received the below error.

```powershell
ErrorRecord                 : The input object cannot be bound to any parameters for the command either because the command does not take pipeline input or the input and
                              its properties do not match any of the parameters that take pipeline input.
WasThrownFromThrowStatement : False
TargetSite                  : System.Collections.ObjectModel.Collection`1[System.Management.Automation.PSObject] Invoke(System.Collections.IEnumerable)
Message                     : The running command stopped because the preference variable "ErrorActionPreference" or common parameter is set to Stop: The input object
                              cannot be bound to any parameters for the command either because the command does not take pipeline input or the input and its properties
                              do not match any of the parameters that take pipeline input.
Data                        : {System.Management.Automation.Interpreter.InterpretedFrameInfo}
InnerException              :
HelpLink                    :
Source                      : System.Management.Automation
HResult                     : -2146233087
StackTrace                  :    at System.Management.Automation.Runspaces.PipelineBase.Invoke(IEnumerable input)
                                 at Microsoft.PowerShell.Executor.ExecuteCommandHelper(Pipeline tempPipeline, Exception& exceptionThrown, ExecutionOptions options)
```

**Comment by gerardog** [comment]
Matt! You antipated me. I will post what I had in draft and then address your comments:

Hi again!
Sorry for not having the time to focus on gsudo as much as I wanted lately. Let's try to move this one forward a little bit.

There is an experimental [`Invoke-gsudo.ps1`](https://github.com/gerardog/gsudo/blob/Feature.PwshSyntax/src/gsudo.extras/Invoke-gsudo.ps1) 
that I want to open for early testing. Put it in your path, preferable in gsudo's folder (`get-command gsudo.exe | % { (get-item $_.Source ).Directory.FullName }`)

This is a glimpse of current status:

``` powershell
.DESCRIPTION
Serializes a scriptblock and executes it in an elevated powershell. 
The ScriptBlock runs in a different process, so it can´t read/write variables from the invoking scope.
If you reference a variable in a scriptblock using the `$using:variableName` it will be replaced with it´s serialized value.
The elevated command can accept input from the pipeline with $Input. It will be serialized, so size matters.
The script result is serialized, sent back to the non-elevated instance, and returned.
Optionally you can check for "$LastExitCode -eq 999" to find out if gsudo failed to elevate (UAC popup cancelled) 

.EXAMPLE
PS> Get-Process notepad | Invoke-gsudo { Stop-Process }

PS> Invoke-gsudo { return Get-Content 'C:\My Secret Folder\My Secret.txt'}

PS> $a=1; $b = Invoke-gsudo { $using:a+10 }; Write-Host "Sum returned: $b";

Sum returned: 11
```

Problems/Challenges

- Steppable input is supported, but serialized all at once. (It's not streamed one-by-one). Large input will probably cause issues.
I don't think this is such a big issue at this point since the user could move the elevation boundary to minimize the serialization and performance hit.

- Commands `return` issue:
As of now, Invoke-gsudo is requiring an extra 'return' statement on some cases. For example:

- `PS C:\> Invoke-Command { Get-Content C:\Test.txt }` for reference, Invoke-Command works (returns file content).
- `PS C:\> Invoke-gsudo { Get-Content C:\Test.txt }` doesn't work, returns nothing.
- `PS C:\> Invoke-gsudo { return Get-Content C:\Test.txt }` works.
- `PS C:\> Invoke-gsudo { Get-Content C:\Test.txt ; 1 }` works, returns the content and a 1.
```

- Exception handling: 

I'm not able to distinguish between Terminating (`Invoke-gsudo { throw}`) and non-terminating errors (` Invoke-gsudo { Write-Error 1; Write-Error 2;}`) [here](https://github.com/gerardog/gsudo/blob/e9ad64b/src/gsudo.extras/Invoke-gsudo.ps1#L152) 
and I think the ErrorAction needs to be Stop for the formers so they trigger catch blocks.

- Only works in TokenMode, at least for now.
Basically dont change gsudo settings like: SecurityEnforceUacIsolation or ForceXXXConsole

Awaiting feedback mode: ON

**Comment by gerardog** [comment]
Matt, `$inputObject` allows to do `Get-Process notepad | Invoke-gsudo { Stop-Process }`. You will see it as `$Input` inside the ScriptBlock since it is an automatic variable generated by Pwsh Invoke-Command. 
Great catch it's causing the 'return' issue. Pushing a fix right now.

**Comment by mattcargile** [comment]
>I'm not able to distinguish between Terminating (Invoke-gsudo { throw}) and non-terminating errors ( Invoke-gsudo { Write-Error 1; Write-Error 2;}) [here](https://github.com/gerardog/gsudo/blob/e9ad64b/src/gsudo.extras/Invoke-gsudo.ps1#L152)
and I think the ErrorAction needs to be Stop for the formers so they trigger catch blocks.

I would think we would let the users determine the `ErrorAction` flow within the `[scriptblock]` passed into `Invoke-gsudo.ps1`? I would assume the aforementioned line should avoid declaring an `ErrorAction` and let the users `$ErrorActionPreference` come into play.

As far as telling the difference between the errors, I was reading [this](https://docs.microsoft.com/en-us/powershell/scripting/developer/cmdlet/interpreting-errorrecord-objects?view=powershell-7.2) and [this](https://docs.microsoft.com/en-us/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError?view=powershellsdk-7.0.0#remarks). Maybe we can use one of these:
* `$Error.CategoryInfo.Category`
* `$Error.Exception.WasThrownFromThrowStatement`

`Category` appears to always be `OperationStopped` due to a `System.Management.Automation.PipelineStoppedException` being called by `ThrowTerminatingError(ErrorRecord)`. This is opposed to a non-terminating error which calls `System.Management.Automation.Cmdlet.WriteError`. Not sure if we need to tell a difference between `Write-Error 1` and `Write-Error 1 -ErrorAction Stop`?

I may be confused with the use cases for the error handling. I was playing with the below.

```powershell
Invoke-gsudo.ps1 { write-error 1; write-error 2; 2 }
```

I would expect it to return
```powershell
Invoke-gsudo.ps1: 1
Invoke-gsudo.ps1: 2
2
```

Currently it is returning
```powershell
2
Invoke-gsudo.ps1: 1
```

I guess it is because of the redirection ( [`2>&1`](https://github.com/gerardog/gsudo/blob/24ad58f7accfcb18b839f5a2a83c218bbcbb87c2/src/gsudo.extras/Invoke-gsudo.ps1#L145) ) changing the order of the errors and then the `ErrorAction` on the aforementioned line.

As another example, I was playing with the below which does work as I would expect.

```powershell
Invoke-gsudo.ps1 { throw 'hello world'; 2 }
```

**Comment by gerardog** [comment]
> > I'm not able to distinguish between Terminating (Invoke-gsudo { throw}) and non-terminating errors ( Invoke-gsudo { Write-Error 1; Write-Error 2;}) [here](https://github.com/gerardog/gsudo/blob/e9ad64b/src/gsudo.extras/Invoke-gsudo.ps1#L152)
> > and I think the ErrorAction needs to be Stop for the formers so they trigger catch blocks.

TIL: Apparently, nobody can. See https://github.com/PowerShell/PowerShell/issues/3158

> I would think we would let the users determine the `ErrorAction` flow within the `[scriptblock]` passed into `Invoke-gsudo.ps1`? I would assume the aforementioned line should avoid declaring an `ErrorAction` and let the users `$ErrorActionPreference` come into play.
> 
I agree. I removed specifying the fixed ErrorAction you are referring to. 

> * `$Error.Exception.WasThrownFromThrowStatement`

Thanks, that really helped. Now I do try/catch on the scriptblock to detect the terminating errors, and force `throw $_` to ensure `WasThrownFromThrowStatement` is true. Works 99% 🤷‍♂️. 

Now, should I honor `Invoke-gsudo -ErrorAction` parameter? Invoke-Command does not. No difference between:
	- `Invoke-Command { "/", "/nofound", "/" | Get-Item } -ErrorAction Stop`
	- `Invoke-Command { "/", "/nofound", "/" | Get-Item }`

It felt harmless to take the `-ErrorAction` param (IF specified) and setting that as `$ErrorActionPreference` in the elevated instance. Opinions?

If omitted the invokers ErrorActionPreference is forwarded to the elevated instance.

> I may be confused with the use cases for the error handling. 

Me too. I'm learning as I code here. Found this great recap on error handling on PS: See https://github.com/MicrosoftDocs/PowerShell-Docs/issues/1583
I'd mimic `Invoke-Command` assuming it's doing the proper thing, or what the user would expect, plus that is easier to communicate. 

In a nutshell, keep the same exception type, keep the same stream. For example:

- Non-terminating errors: 
  - Should not terminate the pipeline execution.
  - Capturable with `-ErrVarible` or `2>&1`
  - E.g. `Invoke-gsudo { "\", "notExisting" | Get-Item } } -errVariable failed`
- Terminating errors
  - Capturable with try/catch
  - `try { Invoke-Gsudo { throw; } } catch {"Catched: $_"}`

Now, I still have problems between **Script-terminating errors** and **Statement-terminating errors**.
Apparently surrounding the script in a try/catch turns a Statement-terminating error into a Script-terminating error (in this context, actually into a scriptblock-terminating).

For example, Invoke-Command behaves differently surrounded by try/catch:

```
PS C:\Users\gerar> Invoke-Command { Some-InvalidCmdLet ; 10 }
Some-InvalidCmdLet: The term 'Some-InvalidCmdLet' is not recognized as a name of a cmdlet, function, script file, or executable program.
10

PS C:\Users\gerar> try { Invoke-Command { Some-InvalidCmdLet ; 10 } } catch { "Catched: $_" }
Catched: The term 'Some-InvalidCmdLet' is not recognized as a name of a cmdlet, function, script file, or executable program.
```
The later stops execution at `Some-InvalidCmdLet` and never returns `10`. 
The problem here is that I use try catch to detect termining errors. 

So, is it correct to conclude that all statement-terminating errors will turn into script-terminating?

> I was playing with the below.
> 
> ```powershell
> Invoke-gsudo.ps1 { write-error 1; write-error 2; 2 }
> ```
> 
> I would expect it to return
> 
> ```powershell
> Invoke-gsudo.ps1: 1
> Invoke-gsudo.ps1: 2
> 2
> ```
> 
> Currently it is returning
> 
> ```powershell
> 2
> Invoke-gsudo.ps1: 1
> ```
> 
> I guess it is because of the redirection ( [`2>&1`](https://github.com/gerardog/gsudo/blob/24ad58f7accfcb18b839f5a2a83c218bbcbb87c2/src/gsudo.extras/Invoke-gsudo.ps1#L145) ) changing the order of the errors and then the `ErrorAction` on the aforementioned line.

Yes, the redirection changes the output order. No workaround AFAIK. 
Also, the second exception was lost. Should be fixed now.

Here are a few test cases. Ideally each pair of lines should behave the same, but they don't. (Invoke-Command vs Invoke-gsudo). 

Proper PS tests may exists in the future, for now a comment will do.

```
try { invoke-gsudo { 0/0 } } catch { "catched: $_" }
try { Invoke-Command { 0/0 } } catch { "catched: $_" }

Invoke-gsudo { 0/0;0/0 } # fail: only 1 exception, caused by try/catch
Invoke-Command { 0/0;0/0 } 

try { invoke-gsudo   { "asd" | ConvertFrom-Json } } catch { "catched: $_" } 
try { Invoke-Command { "asd" | ConvertFrom-Json } } catch { "catched: $_" } # Weird, why catched? Oh.. PowerShell/PowerShell#2860

try { Invoke-Command { throw } } catch { "catched: $_" }
try { invoke-gsudo   { throw } } catch { "catched: $_" }

try { invoke-Command { "notfound" | Get-Item } } catch { "catched: $_" }
try { invoke-gsudo   { "notfound" | Get-Item } } catch { "catched: $_" }

try { invoke-Command { "notfound" | Get-Item -ErrorAction Stop } } catch { "catched: $_" }
try { invoke-gsudo   { "notfound" | Get-Item -ErrorAction Stop } } catch { "catched: $_" }

try { invoke-Command { ".", "notfound" | Get-Item -ErrorAction Stop } } catch { "catched: $_" }
try { invoke-gsudo   { ".", "notfound" | Get-Item -ErrorAction Stop } } catch { "catched: $_" }

#weird fail. Also, run as admin
try { invoke-command { Get-Service "netlogon" | Suspend-Service -ErrorAction Stop }  } catch { "catched: $_" }
try { invoke-gsudo { Get-Service "netlogon" | Suspend-Service -ErrorAction Stop }  } catch { "catched: $_" }

try { invoke-Command { Get-InvalidCmdLet ; 1 } } catch { "catched: $_" }
try { invoke-gsudo   { Get-InvalidCmdLet ; 1 } } catch { "catched: $_" }

try { invoke-gsudo { [int]::Parse('foo') }  } catch { "e $_"  }
try { invoke-command { [int]::Parse('foo') }  } catch { "e $_"  }

```

I've just pushed a [new](https://github.com/gerardog/gsudo/blob/Feature.PwshSyntax/src/gsudo.extras/Invoke-gsudo.ps1) version. Feedback welcomed!

**Comment by gerardog** [comment]
Uploaded tests, and bugfixing.
```
PS C:\git\gsudo> Invoke-Pester -Output Detailed
Pester v5.3.1

Starting discovery in 2 files.
Discovery found 16 tests in 205ms.
Running tests.

Running tests from 'C:\git\gsudo\src\gsudo.extras\gsudo.Tests.ps1'
Describing PS Gsudo (v7.2.1)
  [+] It serializes return values as string. 2.33s (2.3s|38ms)
  [+] When invoked as gsudo !!, It elevates the last command executed 452ms (451ms|1ms)

Running tests from 'C:\git\gsudo\src\gsudo.extras\Invoke-gsudo.Tests.ps1'
Describing PS Invoke-Gsudo (v7.2.1)
  [+] It serializes return values maintaining its type 569ms (568ms|1ms)
  [+] It serializes return values mantaining its properties. 503ms (502ms|1ms)
  [+] It returns an array of values mantaining its properties. 524ms (523ms|1ms)
  [+] It accepts objects from the pipeline. 482ms (480ms|2ms)
  [+] It throws when Error thrown 622ms (621ms|1ms)
  [+] It throws with expression runtime errors 1.14s (1.14s|1ms)
  [+] It throws with .Net Exceptions 577ms (577ms|1ms)
  [+] It throws when ErrorAction = Stop 570ms (570ms|1ms)
  [+] It throws when ErrorActionPreference = Stop 550ms (550ms|0ms)
  [+] It forwards ErrorActionPreference 'Stop' to the elevated instance 527ms (525ms|2ms)
  [+] It forwards ErrorActionPreference 'Continue' to the elevated instance 498ms (498ms|1ms)
  [+] It forwards ErrorActionPreference 'Ignore' to the elevated instance 494ms (493ms|1ms)
  [+] It doesn't throw when ErrorActionPreference = Continue 1.23s (1.23s|1ms)
  [+] It doesn't throw with '-ErrorAction Continue-' 2.69s (2.69s|1ms)
Tests completed in 49.29s
Tests Passed: 16, Failed: 0, Skipped: 0 NotRun: 0
```
Also on the build server: [CI test results](https://ci.appveyor.com/project/gerardog/gsudo/builds/42436518/tests) with inconvenient ordering.

And last, but not least: I added a gsudo PowerShell Module: (Add `Import-Module pathto/gsudoModule.psm1` to your `$PROFILE`) which adds support for `gsudo !!` too. (As requested in #44)

**Comment by gerardog** [comment]
Released in [v1.1.0](https://github.com/gerardog/gsudo/releases/tag/v1.1.0)
